front50-proto/src/main/proto/applications.proto

```diff
@@ -0,0 +1,241 @@
+  google.protobuf.Struct data = 2;
+
+  // The required permissions for updating an attribute group.
+  repeated Permission permissions = 3;
+}
+
+// InstanceLinkGroup is a titled collection of InstanceLinks.
+message InstanceLinkGroup {
+  // The title of the group.
+  string title = 1;
+
+  // A collection of instance links.
+  repeated InstanceLink links = 2;
+}
+
+// InstanceLink is a titled URL link template for server group instances.
+message InstanceLink {
+  // The title of an instance link.
+  string title = 1;
+
+  // The link template value.
+  string link = 2;
+}
+
+// Notification is the configuration for application-wide notification recipients.
+message Notification {
+  // The type of notification (e.g. slack, email, webhook...)
+  string notification_type = 1;
+
+  // The type-specific notification recipient (e.g. email address, slack channel(s), webhook endpoint...)
+  string recipient = 2;
+
+  // A collection of event names to notify on.
+  repeated string notifying_events = 3;
+
+  // A bucket of type-specific notification config.
+  map<string, string> metadata = 6;
+}
+
+// Permission defines an authorization and its allowed prinicpals.
+message Permission {
+  // The name of the authorization (e.g. WRITE, READ, EXECUTE...)
+  string name = 1;
+
+  // A collection of principals who are allowed this permission.
+  repeated string allowed_principals = 2;
+}
+
+// ListApplicationsRequest can be used to list and search applications known to Spinnaker.
+message ListApplicationsRequest {
+  // The number of applications to return in the response.
+  //
+  // Defaults to 100.
+  int32 page_size = 1;
+
+  // The page token to continue pagination from. If left unset, the first `page_size` results will be returned.
+  string page_token = 2;
+
+  // Allows filtering of applications by fields.
+  string filter = 3;
+
+  // Order applications by fields and direction.
+  string order_by = 4;
+
+  // Define partial responses of an application.
+  google.protobuf.FieldMask fields = 5;
+}
+
+// ListApplicationsResponse returns a page of applications matching the provided request.
+message ListApplicationsResponse {
+  // The applications returned.
+  repeated Application applications = 1;
+
+  // The next page token, if any.
+  string next_page_token = 3;
+}
+
+// GetApplicationRequest returns a single application.
+message GetApplicationRequest {
+  // Required. Resource name for the application (e.g. "applications/front50")
+  string name = 1;
+
+  // Define partial responses of an application.
+  google.protobuf.FieldMask fields = 2;
+}
+
+// CreateApplicationRequest will attempt to create an application.
+message CreateApplicationRequest {
+  // Required. The application to be created.
+  Application application = 1;
+
+  // If true, it indicates that the given request should only be validated, not executed.
+  bool validate_only = 2;
+
+  // A unique string id used for detecting duplicated requests.
+  string request_id = 3;
```

Comment:
How is this going to work? Does the created/updated application contain the `last_request_id` and if we see `request_id==last_request_id` we return success?

Reply:
I wasn't planning on requiring an optimistic locking model, instead having a client-generated hash that would be used strictly for de-duplication. Certainly open to feedback here, I definitely see the benefits of what you're proposing.

Reply:
Just in case the client accidentally sent the same request twice, basically? But what if those requests got load-balanced to different front50 instances or something... it seems like a hard problem to solve well, so I'd lean in favor of not trying to solve it at all, but maybe I'm just being difficult.

Reply:
OK, that's legit. I'll keep `request_id` usages isolated to long polling operations - which won't be part of this initial go-around. Thanks!